### PR TITLE
Improve parallelMap implementation by removing iteration for force unwrapping every element

### PR DESCRIPTION
**Pull request by cltnschlosser**
#3515 Wouldn't reopen, so created new PR

### CODE REVIEW

File: Source/swiftlint/Extensions/Array+SwiftLint.swift [hunk 1]
```diff
@@ -6,12 +6,14 @@ extension Array {
     }
 
     func parallelMap<T>(transform: (Element) -> T) -> [T] {
-        var result = ContiguousArray<T?>(repeating: nil, count: count)
-        return result.withUnsafeMutableBufferPointer { buffer in
-            DispatchQueue.concurrentPerform(iterations: buffer.count) { idx in
-                buffer[idx] = transform(self[idx])
+        return [T](unsafeUninitializedCapacity: count) { buffer, initializedCount in
+            let baseAddress = buffer.baseAddress!
```
Comment (jpsim) [inline Source/swiftlint/Extensions/Array+SwiftLint.swift:10]: Is this always guaranteed to succeed? Is that behavior documented?

File: Source/swiftlint/Extensions/Array+SwiftLint.swift [hunk 2]
```diff
@@ -6,12 +6,14 @@ extension Array {
...
+            DispatchQueue.concurrentPerform(iterations: count) { index in
+                // Using buffer[index] does assignWithTake which tries
+                // to read the uninitialized value (to release it) and crashes
+                (baseAddress + index).initialize(to: transform(self[index]))
```
Comment (jpsim) [inline Source/swiftlint/Extensions/Array+SwiftLint.swift:14]: How does this work? I’d expect to have to multiply the index by the size of the transformed result type.

Comment (cltnschlosser) [reply Source/swiftlint/Extensions/Array+SwiftLint.swift:14]: It used the type of the pointer to know how much to increase by. The `+` operator isn't documented, but it looks like this new function is and is documented to behave the same way: https://developer.apple.com/documentation/swift/unsafemutablepointer/3393143-advanced

File: Source/swiftlint/Extensions/Array+SwiftLint.swift (hunk 1 above)
Comment (cltnschlosser) [reply Source/swiftlint/Extensions/Array+SwiftLint.swift:10]: In this case yes. Documented, not really. That's how I ran into this: https://bugs.swift.org/browse/SR-14164

If you follow throw this, the buffer is initialized with a non-optional baseAddress: https://github.com/apple/swift/blob/06dc593cece2ded7457ba3d5a6c41857881da157/stdlib/public/core/Array.swift#L1468

**Review by jpsim — APPROVED** [review]
Thanks for answering my questions. Looks better than the previous implementation 👌

**Pull request merged**